### PR TITLE
Filter and export App instance logs to app specific log API using App…

### PR DESCRIPTION
**Pull request by gkodali-zededa**
… specific proto buf for logs

Signed-off-by: GopiKrishna Kodali <gkodali@zededa.com>

### CODE REVIEW

File: pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch [hunk 1]
```diff
@@ -39,7 +39,7 @@ index a43c57edad..d0bf636c10 100644
  
  	snprintf(logfile, PATH_MAX-1, "%s/guest-%s%s.log",
  		 log_dir, data, con->log_suffix);
-+        snprintf(logname, PATH_MAX-1, "guest_vm/%s%s",
++        snprintf(logname, PATH_MAX-1, "guest_vm-%s%s",
```
Comment (rvs) [inline pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch]: I'm just curious - why this change?

Comment (eriknordmark) [reply pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch]: I guess it needs to be a filename with no '/ and not a pathname.

Comment (rvs) [reply pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch]: I must be missing something, but looking at the rest of the code -- it simply looks for 'guest_vmX' when it parses that field (for any value of X really)

What am I missing?

Comment (gkodali-zededa) [reply pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch]: Except for this case, the other xen tools logs like xl, qemu-dm etc use "-" instead of "/". Wanted everything to look same.

Comment (rvs) [reply pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch]: Ah! That makes sense then.

Comment (rvs) [reply pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch]: Btw, please don't forget to update the documentation tho @gkodali-zededa

Comment (gkodali-zededa) [reply pkg/xen-tools/patches-4.13.0/08-memlogd-logger.patch]: will change LOGGING.md to reflect this change.

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 2]
```diff
@@ -536,81 +532,99 @@ func processEvents(image string, prevLastSent time.Time,
...
-	filename := fmt.Sprintf("%s/%s/%s", types.PersistDir, dirname, image)
-	_, err := os.Stat(filename)
-	if err != nil {
-		file, err := os.Create(filename)
-		if err != nil {
-			log.Infof("recordLast: %s\n", err)
-			return
+func flushAllLogBundles(image string, iteration int, eveVersion string,
+	reportLogs *logs.LogBundle, appLogBundles map[string]*logs.AppInstanceLogBundle) bool {
+	sent := sendProtoStrForLogs(reportLogs, image, iteration, eveVersion)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go]: There could be zero events in this bundle, Should avoid sending in that case.

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 3]
```diff
@@ -536,81 +532,99 @@ func processEvents(image string, prevLastSent time.Time,
...
+	if !sent {
+		return false
+	}
+	appBundlesToDelete := []string{}
+	for appUUID, appLogBundle := range appLogBundles {
+		log.Debugf("flushAllLogBundles: Trying to flush App bundle with UUID %s and %d logs",
+			appUUID, len(appLogBundle.Log))
+		sent = sendProtoStrForAppLogs(appUUID, appLogBundle, iteration, image)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go]: There could be zero events in this bundle, Should avoid sending in that case.

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 4]
```diff
@@ -536,81 +532,99 @@ func processEvents(image string, prevLastSent time.Time,
...
+		delete(appLogBundles, appUUID)
 	}
-	now := time.Now()
-	err = os.Chtimes(filename, now, now)
-	if err != nil {
-		log.Errorf("recordLast: %s\n", err)
+	return sent
+}
+
+var msgIDCounter = 1
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go]: Don't you want this per app? Don't know how it gets used on the server side but it would potentially be confusing to see gaps just because there are multiple app instances.

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 5]
```diff
@@ -651,6 +665,9 @@ func handleLogEvent(event logEntry, reportLogs *logs.LogBundle, counter int) {
 // Returns true if a message was successfully sent
 func sendProtoStrForLogs(reportLogs *logs.LogBundle, image string,
 	iteration int, eveVersion string) bool {
+	if len(reportLogs.Log) == 0 {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go]: Ah - you have the check here.

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 2 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: Taken care of inside the callee

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 3 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: Taken care of inside the callee

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 4 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: I've brought this up in meetings earlier about the relevance of this field while ordering messages. Turns out nobody seems to care and know how this can be used. BTW if the software tool itself can order ascending/descending fashion, gaps should not make any difference. If it's not used by consumer (cloud), it makes no sense to put any more complexity in device IMHO.

Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: @nordmark you are right. Please forgive my ignorance. Yes, even if they are being channeled to different endpoints, cloud will see gaps in message ids. But, if cloud can maintain the order, we should be good,

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 5 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: yep

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 6]
```diff
@@ -463,12 +447,11 @@ func handleDNSDelete(ctxArg interface{}, key string, statusArg interface{}) {
...
 // Compares and drops events which have already been sent to the cloud
-func processEvents(image string, prevLastSent time.Time,
-	logChan <-chan logEntry, eveVersion string) {
-
-	log.Infof("processEvents(%s, %s)\n", image, prevLastSent.String())
+func processEvents(image string, logChan <-chan logEntry,
+	eveVersion string, ctx *logmanagerContext) {
 
 	reportLogs := new(logs.LogBundle)
+	appLogBundles := make(map[string]*logs.AppInstanceLogBundle)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go]: How about having the right had side of this be a struct so you can include the msgID?

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 4 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/logmanager/logmanager.go]: Including it with gaps is just confusiong; either it shouldn't be there or it should have some useful meaning
See above for suggestion

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 6 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: Will do.

Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: That might not work. We delete bundles that are exported to cloud. We might have to implement a separate map for maintaining per-app counters. But that brings in more complexity where we have to keep track of newly added/deleted apps.

Comment (eriknordmark) [reply pkg/pillar/cmd/logmanager/logmanager.go]: If we can't make the number make sense I'd prefer setting it to zero then including something which intentionally has an undeterminable amount of gaps in the sequence. Those gaps will be utterly confusing.

Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: setting msgId to zero for app logs for now.

Comment (eriknordmark) [reply pkg/pillar/cmd/logmanager/logmanager.go]: OK, once I see that code removed from the PR I'll pull this.

Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: It is already removed.

Comment (eriknordmark) [reply pkg/pillar/cmd/logmanager/logmanager.go]: Now it is; you hadn't pushed that an hour ago.

Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go]: I pushed the changes 9 hours ago. Not sure why you did not see them in PR.
